airflow/sensors/smart_sensor_operator.py

```diff
@@ -0,0 +1,730 @@
+        # Join with DagRun table will be very slow based on the number of sensor tasks we
+        # need to handle. We query all smart tasks in this operator
+        # and expect scheduler correct the states in _change_state_for_tis_without_dagrun()
+
+        sensor_works = []
+        for ti in tis:
+            try:
+                sensor_works.append(SensorWork(ti))
+            except Exception as e:  # pylint: disable=broad-except
+                self.log.exception("Exception at creating sensor work for ti %s", ti.key)
+                self.log.exception(e, exc_info=True)
+
+        self.log.info("%d tasks detected.", len(sensor_works))
+
+        new_sensor_works = [x for x in sensor_works if x not in self.sensor_works]
+
+        self._update_ti_hostname(new_sensor_works)
+
+        self.sensor_works = sensor_works
+
+    @provide_session
+    def _update_ti_hostname(self, sensor_works, session=None):
+        """
+        Update task instance hostname for new sensor works.
+        :param sensor_works:
+        :return:
+        """
+        TI = TaskInstance
+        ti_keys = [(x.dag_id, x.task_id, x.execution_date) for x in sensor_works]
+
+        def update_ti_hostname_with_count(count, ti_keys):
+            # Using or_ instead of in_ here to prevent from full table scan.
+            tis = session.query(TI) \
+                .filter(or_(tuple_(TI.dag_id, TI.task_id, TI.execution_date) == ti_key
+                            for ti_key in ti_keys)) \
+                .all()
+
+            for ti in tis:
+                ti.hostname = self.hostname
+            session.commit()
+
+            return count + len(ti_keys)
+
+        count = helpers.reduce_in_chunks(update_ti_hostname_with_count, ti_keys, 0, self.max_tis_per_query)
+        if count:
+            self.log.info("Updated hostname on %s tis.", count)
+
+    @provide_session
+    def _mark_multi_state(self, operator, poke_hash, encoded_poke_context, state, session=None):
+        """
+        Mark state for multiple tasks that have hashcode=poke_hash.
+        :param poke_hash:
+        :param state:
+        :param session:
+        :return:
+        """
+
+        def mark_state(ti, sensor_instance):
+            ti.state = state
+            sensor_instance.state = state
+            if state in State.finished():
+                ti.end_date = end_date
+                ti.set_duration()
+
+        SI = SensorInstance
+        TI = TaskInstance
+
+        count_marked = 0
+        try:
+            query_result = session.query(TI, SI)\
+                .join(TI, and_(TI.dag_id == SI.dag_id,
+                               TI.task_id == SI.task_id,
+                               TI.execution_date == SI.execution_date)) \
+                .filter(SI.state == State.SENSING) \
+                .filter(SI.hashcode == poke_hash) \
+                .filter(SI.operator == operator) \
+                .with_for_update().all()
```

Comment:
I am concerned about any join that involves TaskInstance as that table can contain large number of records.

Did you find any issues with this with your Prod cluster?

Reply:
How about if we filter first with `State.SENSING` to decrease the set we do joins, WDYT?

Reply:
Thanks Kaxil for raising this. Yes the query performance is really important. I think this query should be safe at least in mysql engine. 

We have the index on the hashcode in sensor_instace table which can largely reduce the time of this query (and also another query for load sensor tasks).

In our experience of using smart sensor, this query did not cause any issue and the max duplicate jobs in our db is around 500 at the same time. We all know that our scale of airflow cluster is much larger than most use cases so I assume if this query is safe in our cluster it seems not likely to cause a performance issue in other system. 

Also, before creating this PR, we have tested with `explain` the queries with different order in our mysql db. For this query the results are the same. DB engine do the optimization and putting the `filter` before or after the `join` in sqlalchemy does not really matter.